docker/archive/writer.go

```diff
@@ -63,6 +79,19 @@ func (w *Writer) Close() error {
 	if err2 := w.writer.Close(); err2 != nil && err == nil {
 		err = err2
 	}
+	if err == nil && w.regularFile && !w.hadCommit {
+		// Writing to the destination never had a success; delete the destination if we created it.
+		// This is done primarily because we don’t implement adding another image to a pre-existing image, so if we
+		// left a partial archive around (notably because reading from the _source_ has failed), we couldn’t retry without
+		// the caller manually deleting the partial archive. So, delete it instead.
+		//
+		// Archives with at least one successfully created image are left around; they might still be valuable.
+		//
+		// Note a corner case: If there _originally_ was an empty file (which is not a valid archive anyway), this deletes it.
+		if err2 := os.Remove(w.path); err2 != nil {
```

Comment:
Why create err2?  Why not just return os.Remove(w.path)? Or set it to 
if err = os.Remote ...

Reply:
I’m actually a bit struggling with the logic of which errors should or should not cause this `Remove` … e.g. if writing the final metadata stream failed, we might still want to remove the unusable archive, but report primarily the metadata write failure back to the user.

I suppose I should just do the extra bit and use a temporary file with `Rename` here, either way…

Reply:
I am much in favor in naming errors rather than enumerating them in such cases. `$subjectError` is much easier for me to brain parse than `err$Number`.